File: matlab/hashFigureArtifacts.m
function manifest = hashFigureArtifacts(outputDir, stem)
%HASHFIGUREARTIFACTS Record provenance and deterministic visual hashes.
arguments
    outputDir (1,1) string
    stem (1,1) string
end

pdfPath = fullfile(outputDir, stem + ".pdf");
svgPath = fullfile(outputDir, stem + ".svg");
pngPath = fullfile(outputDir, stem + ".png");
required = [pdfPath, svgPath, pngPath];
for path = required
    if ~isfile(path)
        error('mathmodeling:MissingFigureArtifact', 'Missing figure artifact: %s', path);
    end
end

svgText = fileread(svgPath);
svgText = regexprep(svgText, '\r\n?', '\n');
svgText = regexprep(svgText, '(?s)<!--.*?-->', '');
svgText = regexprep(svgText, '(?is)<metadata\b.*?</metadata>', '');
svgText = regexprep(svgText, 'fillPattern[0-9]+_', 'fillPattern_');
svgText = regexprep(svgText, '[ \t]+(?=\n|$)', '');
svgCanonical = strtrim(svgText);

[pngImage, colorMap, alpha] = imread(pngPath);
if ~isempty(colorMap)
    rgb = uint8(round(255 * ind2rgb(pngImage, colorMap)));
elseif ndims(pngImage) == 2
    gray = mmToUint8(pngImage);
    rgb = repmat(gray, [1, 1, 3]);
else
    rgb = mmToUint8(pngImage(:, :, 1:3));
    if size(pngImage, 3) >= 4 && isempty(alpha)
        alpha = pngImage(:, :, 4);
    end
end
if isempty(alpha)
    alpha = uint8(255 * ones(size(rgb, 1), size(rgb, 2)));
else
    alpha = mmToUint8(alpha);
end
rgba = cat(3, rgb, alpha);
height = size(rgba, 1);
width = size(rgba, 2);
pixelHeader = unicode2native(sprintf('RGBA|%dx%d|uint8|', width, height), 'UTF-8');
pixelBytes = reshape(permute(rgba, [3, 2, 1]), 1, []);

manifest = struct();
manifest.schema_version = 1;
manifest.stem = char(stem);
manifest.artifacts = struct();
manifest.artifacts.pdf = mmRawRecord(pdfPath);
manifest.artifacts.pdf.determinism_role = 'provenance-only';
manifest.artifacts.svg = mmRawRecord(svgPath);
manifest.artifacts.svg.canonical_sha256 = mmSha256(unicode2native(svgCanonical, 'UTF-8'));
manifest.artifacts.svg.canonicalization = 'utf8-lf-no-comments-no-metadata-fillpattern-normalized-rstrip';
manifest.artifacts.png = mmRawRecord(pngPath);
manifest.artifacts.png.pixel_sha256 = mmSha256([pixelHeader, pixelBytes]);
manifest.artifacts.png.pixel_encoding = sprintf('RGBA|%dx%d|uint8|row-major', width, height);
manifest.reproducibility_note = [ ...
    'PDF raw SHA-256 is retained for provenance, but export metadata may change raw bytes. ', ...
    'Use canonical SVG SHA-256 and decoded PNG pixel SHA-256 as deterministic visual checks.' ...
];

manifestPath = fullfile(outputDir, stem + ".hashes.json");
fileId = fopen(manifestPath, 'w', 'n', 'UTF-8');
assert(fileId >= 0, 'Unable to write hash manifest: %s', manifestPath);
cleanup = onCleanup(@() fclose(fileId));
fwrite(fileId, jsonencode(manifest, PrettyPrint=true), 'char');
end

function record = mmRawRecord(path)
info = dir(path);
record = struct('path', char(path), 'bytes', info.bytes, ...
    'raw_sha256', mmSha256(mmReadBytes(path)));
end

function bytes = mmReadBytes(path)
fileId = fopen(path, 'rb');
assert(fileId >= 0, 'Unable to read artifact: %s', path);
cleanup = onCleanup(@() fclose(fileId));
bytes = fread(fileId, Inf, '*uint8').';
end

function hex = mmSha256(bytes)
digest = java.security.MessageDigest.getInstance('SHA-256');
digest.update(typecast(uint8(bytes(:)), 'int8'));
raw = typecast(digest.digest(), 'uint8');
hex = lower(reshape(dec2hex(raw, 2).', 1, []));
end

function value = mmToUint8(value)
if isa(value, 'uint8')
    return;
elseif isa(value, 'uint16')
    value = uint8(round(double(value) / 257));
elseif islogical(value)
    value = uint8(value) * 255;
else
    value = uint8(round(255 * min(max(double(value), 0), 1)));
end
end

File: src/utils/hash_figure_artifacts.py
from __future__ import annotations

import argparse
import hashlib
import json
import re
from pathlib import Path


def raw_sha256(path: Path) -> str:
    digest = hashlib.sha256()
    with path.open("rb") as handle:
        for chunk in iter(lambda: handle.read(1024 * 1024), b""):
            digest.update(chunk)
    return digest.hexdigest()


def canonical_svg_bytes(path: Path) -> bytes:
    text = path.read_text(encoding="utf-8-sig")
    text = text.replace("\r\n", "\n").replace("\r", "\n")
    text = re.sub(r"<!--.*?-->", "", text, flags=re.DOTALL)
    text = re.sub(r"<metadata\b.*?</metadata>", "", text, flags=re.DOTALL | re.IGNORECASE)
    text = re.sub(r"fillPattern[0-9]+_", "fillPattern_", text)
    text = "\n".join(line.rstrip(" \t") for line in text.split("\n")).strip()
    return text.encode("utf-8")


def png_pixel_payload(path: Path) -> tuple[bytes, int, int]:
    from PIL import Image

    with Image.open(path) as image:
        rgba = image.convert("RGBA")
        width, height = rgba.size
        header = f"RGBA|{width}x{height}|uint8|".encode("ascii")
        return header + rgba.tobytes(), width, height


def build_manifest(directory: Path, stem: str) -> dict:
    pdf = directory / f"{stem}.pdf"
    svg = directory / f"{stem}.svg"
    png = directory / f"{stem}.png"
    for path in (pdf, svg, png):
        if not path.is_file():
            raise FileNotFoundError(path)

    pixel_payload, width, height = png_pixel_payload(png)
    return {
        "schema_version": 1,
        "stem": stem,
        "artifacts": {
            "pdf": {
                "path": str(pdf),
                "bytes": pdf.stat().st_size,
                "raw_sha256": raw_sha256(pdf),
                "determinism_role": "provenance-only",
            },
            "svg": {
                "path": str(svg),
                "bytes": svg.stat().st_size,
                "raw_sha256": raw_sha256(svg),
                "canonical_sha256": hashlib.sha256(canonical_svg_bytes(svg)).hexdigest(),
                "canonicalization": "utf8-lf-no-comments-no-metadata-fillpattern-normalized-rstrip",
            },
            "png": {
                "path": str(png),
                "bytes": png.stat().st_size,
                "raw_sha256": raw_sha256(png),
                "pixel_sha256": hashlib.sha256(pixel_payload).hexdigest(),
                "pixel_encoding": f"RGBA|{width}x{height}|uint8|row-major",
            },
        },
        "reproducibility_note": (
            "PDF raw SHA-256 is retained for provenance, but export metadata may change raw bytes. "
            "Use canonical SVG SHA-256 and decoded PNG pixel SHA-256 as deterministic visual checks."
        ),
    }


def main() -> int:
    parser = argparse.ArgumentParser(description="Hash a MATLAB PDF/SVG/PNG figure triplet.")
    parser.add_argument("--directory", type=Path, required=True)
    parser.add_argument("--stem", required=True)
    parser.add_argument("--output", type=Path)
    args = parser.parse_args()
    manifest = build_manifest(args.directory.resolve(), args.stem)
    output = args.output or args.directory / f"{args.stem}.hashes.json"
    output.parent.mkdir(parents=True, exist_ok=True)
    temporary = output.with_name(f"{output.name}.tmp")
    temporary.write_text(json.dumps(manifest, ensure_ascii=False, indent=2), encoding="utf-8")
    temporary.replace(output)
    print(json.dumps(manifest, ensure_ascii=False, indent=2))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
